Zend/zend_opcode.c

```diff
@@ -153,6 +153,12 @@ ZEND_API void zend_function_dtor(zval *zv)
 		ZEND_ASSERT(function->common.function_name);
 		zend_string_release_ex(function->common.function_name, 1);
 
+		if (function->common.attributes) {
+			zend_hash_destroy(function->common.attributes);
```

Comment:
Looking again, it looks like attributes on functions are freed in `ZEND_API void destroy_op_array(zend_op_array *op_array)`, which already calls zend_hash_destroy.

So you just need the `free` to free attribute tables for global functions.

Reply:
If I comment out `zend_hash_destroy(function->common.attributes);`, then:

```
<?php

var_dump(zend_test_parameter_with_attribute("foo"));

?>
```

will still leak memory:

```
$ valgrind --leak-check=full sapi/cli/php test4.php
==530513== Memcheck, a memory error detector
==530513== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==530513== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==530513== Command: sapi/cli/php test4.php
==530513== 
bool(true)
==530513== 
==530513== HEAP SUMMARY:
==530513==     in use at exit: 280 bytes in 4 blocks
==530513==   total heap usage: 11,608 allocs, 11,604 frees, 1,707,398 bytes allocated
==530513== 
==530513== 280 (56 direct, 224 indirect) bytes in 1 blocks are definitely lost in loss record 4 of 4
==530513==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==530513==    by 0x6D9FA2: __zend_malloc (zend_alloc.c:3068)
==530513==    by 0x745138: zend_add_attribute (zend_attributes.c:226)
==530513==    by 0x6680D1: zend_add_parameter_attribute (zend_attributes.h:102)
==530513==    by 0x66B787: zm_startup_zend_test (test.c:478)
==530513==    by 0x7224CD: zend_startup_module_ex (zend_API.c:2202)
==530513==    by 0x72252C: zend_startup_module_zval (zend_API.c:2217)
==530513==    by 0x734288: zend_hash_apply (zend_hash.c:2011)
==530513==    by 0x722C30: zend_startup_modules (zend_API.c:2328)
==530513==    by 0x67409B: php_module_startup (main.c:2256)
==530513==    by 0x88EDDE: php_cli_startup (php_cli.c:409)
==530513==    by 0x890F61: main (php_cli.c:1334)
==530513== 
==530513== LEAK SUMMARY:
==530513==    definitely lost: 56 bytes in 1 blocks
==530513==    indirectly lost: 224 bytes in 3 blocks
==530513==      possibly lost: 0 bytes in 0 blocks
==530513==    still reachable: 0 bytes in 0 blocks
==530513==         suppressed: 0 bytes in 0 blocks
==530513== 
==530513== For lists of detected and suppressed errors, rerun with: -s
==530513== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```